File: api/routers/mcp.py
"""MCP Client API routes."""

from __future__ import annotations

from typing import List

from fastapi import APIRouter, HTTPException
from loguru import logger

from api import mcp_service
from api.mcp_models import (
    ChatToolCallResponse,
    McpConnectionAuthUpdateRequest,
    McpConnectionCreateRequest,
    McpConnectionResponse,
    McpToolResponse,
)
from construction_os.exceptions import InvalidInputError, NotFoundError
from construction_os.domain import mcp as _mcp_domain  # noqa: F401 — register models

router = APIRouter()


@router.get("/mcp/connections", response_model=List[McpConnectionResponse])
async def list_connections():
    try:
        return await mcp_service.list_connections()
    except Exception as e:
        logger.error(f"Error listing MCP connections: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/mcp/connections", response_model=McpConnectionResponse)
async def create_connection(body: McpConnectionCreateRequest):
    try:
        return await mcp_service.create_connection(body)
    except InvalidInputError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Error creating MCP connection: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/mcp/connections/{connection_id}", response_model=McpConnectionResponse)
async def get_connection(connection_id: str):
    try:
        return await mcp_service.get_connection(connection_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/mcp/connections/{connection_id}")
async def delete_connection(connection_id: str):
    try:
        await mcp_service.delete_connection(connection_id)
        return {"success": True}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put(
    "/mcp/connections/{connection_id}/auth",
    response_model=McpConnectionResponse,
)
async def update_auth(connection_id: str, body: McpConnectionAuthUpdateRequest):
    try:
        return await mcp_service.update_auth(connection_id, body)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except (InvalidInputError, ValueError) as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post(
    "/mcp/connections/{connection_id}/test",
    response_model=McpConnectionResponse,
)
async def test_connection(connection_id: str):
    try:
        return await mcp_service.test_mcp_connection(connection_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post(
    "/mcp/connections/{connection_id}/sync",
    response_model=McpConnectionResponse,
)
async def sync_connection(connection_id: str):
    try:
        return await mcp_service.sync_mcp_connection(connection_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get(
    "/mcp/connections/{connection_id}/tools",
    response_model=List[McpToolResponse],
)
async def list_connection_tools(connection_id: str):
    try:
        return await mcp_service.list_connection_tools(connection_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/mcp/tools/selectable", response_model=List[McpToolResponse])
async def list_selectable_tools():
    try:
        return await mcp_service.list_selectable_tools()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get(
    "/mcp/sessions/{session_id}/tool-calls",
    response_model=List[ChatToolCallResponse],
)
async def list_session_tool_calls(session_id: str):
    try:
        return await mcp_service.list_session_tool_calls(session_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/api/mcp.ts
import apiClient from '@/lib/api/client'
import {
  ChatToolCall,
  CreateMcpConnectionRequest,
  McpConnection,
  McpTool,
  UpdateMcpConnectionAuthRequest,
} from '@/lib/types/mcp'

export const mcpApi = {
  listConnections: async () => {
    const response = await apiClient.get<McpConnection[]>('/mcp/connections')
    return response.data
  },

  getConnection: async (id: string) => {
    const response = await apiClient.get<McpConnection>(`/mcp/connections/${id}`)
    return response.data
  },

  createConnection: async (data: CreateMcpConnectionRequest) => {
    const response = await apiClient.post<McpConnection>('/mcp/connections', data)
    return response.data
  },

  deleteConnection: async (id: string) => {
    await apiClient.delete(`/mcp/connections/${id}`)
  },

  updateAuth: async (id: string, data: UpdateMcpConnectionAuthRequest) => {
    const response = await apiClient.put<McpConnection>(`/mcp/connections/${id}/auth`, data)
    return response.data
  },

  testConnection: async (id: string) => {
    const response = await apiClient.post<McpConnection>(`/mcp/connections/${id}/test`)
    return response.data
  },

  syncConnection: async (id: string) => {
    const response = await apiClient.post<McpConnection>(`/mcp/connections/${id}/sync`)
    return response.data
  },

  listConnectionTools: async (id: string) => {
    const response = await apiClient.get<McpTool[]>(`/mcp/connections/${id}/tools`)
    return response.data
  },

  listSelectableTools: async () => {
    const response = await apiClient.get<McpTool[]>('/mcp/tools/selectable')
    return response.data
  },

  listSessionToolCalls: async (sessionId: string) => {
    const response = await apiClient.get<ChatToolCall[]>(
      `/mcp/sessions/${sessionId}/tool-calls`
    )
    return response.data
  },
}
